diff --git a/drivers/power/supply/qcom/smb5-lib-munch.h b/drivers/power/supply/qcom/smb5-lib-munch.h
--- a/drivers/power/supply/qcom/smb5-lib-munch.h
+++ b/drivers/power/supply/qcom/smb5-lib-munch.h
@@ -99,6 +99,7 @@
 #define BBC_CHARGER_VOTER		"BBC_CHARGER_VOTER"
 #define QC2_UNSUPPORTED_VOTER		"QC2_UNSUPPORTED_VOTER"
 #define THERMAL_FCC_OVERRIDE_VOTER  "THERMAL_FCC_OVERRIDE_VOTER"
+#define BYPASS_VOTER			"BYPASS_VOTER"
 /* use for QC3P5 */
 #define QC3P5_VOTER			"QC3P5_VOTER"
 #define FCC_MAX_QC3P5_VOTER		"FCC_MAX_QC3P5_VOTER"

diff --git a/drivers/power/supply/qcom/smb5-lib-munch.c b/drivers/power/supply/qcom/smb5-lib-munch.c
--- a/drivers/power/supply/qcom/smb5-lib-munch.c
+++ b/drivers/power/supply/qcom/smb5-lib-munch.c
@@ -38,6 +38,7 @@
 	&& (!chg->typec_legacy || chg->typec_legacy_use_rp_icl))
 
 bool off_charge_flag;
 static bool first_boot_flag;
+static int bypass_charging = 0;
 
 static void update_sw_icl_max(struct smb_charger *chg, int pst);
 static int smblib_get_prop_typec_mode(struct smb_charger *chg);
@@ -2408,9 +2409,15 @@ int smblib_get_prop_input_suspend(struct smb_charger *chg,
 int smblib_get_prop_input_suspend(struct smb_charger *chg,
 				  union power_supply_propval *val)
 {
-	val->intval
-		= (get_client_vote(chg->usb_icl_votable, USER_VOTER) == 0)
-			 || get_client_vote(chg->dc_suspend_votable, USER_VOTER);
+	if (get_client_vote(chg->chg_disable_votable, BYPASS_VOTER) == 1)
+		val->intval = 1;
+	else if (bypass_charging)
+		val->intval = 2;
+	else
+		val->intval = 0;
+
 	return 0;
 }
@@ -3173,22 +3180,37 @@ int smblib_set_prop_input_suspend(struct smb_charger *chg,
 int smblib_set_prop_input_suspend(struct smb_charger *chg,
 				  const union power_supply_propval *val)
 {
-	int rc;
+	int rc;
 
 	/* vote 0mA when suspended */
-	rc = vote(chg->usb_icl_votable, USER_VOTER, (bool)val->intval, 0);
+	rc = vote(chg->usb_icl_votable, USER_VOTER, false, 0);
 	if (rc < 0) {
 		smblib_err(chg, "Couldn't vote to %s USB rc=%d\\n",
 			(bool)val->intval ? "suspend" : "resume", rc);
 		return rc;
 	}
 
-	rc = vote(chg->dc_suspend_votable, USER_VOTER, (bool)val->intval, 0);
+	rc = vote(chg->dc_suspend_votable, USER_VOTER, false, 0);
 	if (rc < 0) {
 		smblib_err(chg, "Couldn't vote to %s DC rc=%d\\n",
 			(bool)val->intval ? "suspend" : "resume", rc);
 		return rc;
 	}
 
+	if (val->intval == 1) {
+		rc = vote(chg->chg_disable_votable, BYPASS_VOTER, true, 0);
+		bypass_charging = 0;
+	} else if (val->intval == 2) {
+		rc = vote(chg->chg_disable_votable, BYPASS_VOTER, false, 0);
+		bypass_charging = 1;
+	} else {
+		rc = vote(chg->chg_disable_votable, BYPASS_VOTER, false, 0);
+		bypass_charging = 0;
+	}
+
+	if (rc < 0) {
+		smblib_err(chg, "Couldn't vote to %d input_suspend rc=%d\\n",
+			val->intval, rc);
+		return rc;
+	}
+
 	power_supply_changed(chg->batt_psy);
 	return rc;
 }
@@ -3604,6 +3626,9 @@ static void smblib_thermal_setting_work(struct work_struct *work)
 	struct smb_charger *chg = container_of(work, struct smb_charger,
 			thermal_setting_work.work);
 
+	if (bypass_charging)
+		chg->pps_thermal_level = 0;
+
 	if (chg->pps_thermal_level > chg->system_temp_level) {
 		if (chg->pps_thermal_level - chg->system_temp_level >= 2)
 			chg->pps_thermal_level -= 2;
@@ -3645,6 +3670,9 @@ int smblib_set_prop_system_temp_level(struct smb_charger *chg,
 		return -EINVAL;
 
 	chg->system_temp_level = val->intval;
+	if (bypass_charging)
+		chg->system_temp_level = 0;
 
 	smblib_dbg(chg, PR_OEM, "thermal level:%d, thermal_levels:%d "
 			"chg->system_temp_level:%d, charger_type:%d\\n",
